File: 251004/Asepkov/internal/service/mark-service.go
package service

import (
	"RESTAPI/internal/dto"
	"RESTAPI/internal/entity"
	"RESTAPI/internal/repository"
	"errors"
)

type MarkService struct {
	repo *repository.MarkRepository
}

func NewMarkService(repo *repository.MarkRepository) *MarkService {
	return &MarkService{repo: repo}
}

func (s *MarkService) Create(req dto.MarkRequestTo) (*dto.MarkResponseTo, error) {
	mark := &entity.Mark{
		Name: req.Name,
	}
	err := s.repo.Create(mark)
	if err != nil {
		return nil, err
	}
	return &dto.MarkResponseTo{
		ID:   mark.ID,
		Name: mark.Name,
	}, nil
}

func (s *MarkService) GetById(id int64) (*dto.MarkResponseTo, error) {
	mark, err := s.repo.GetById(id)
	if err != nil {
		return nil, errors.New("mark not found")
	}
	return &dto.MarkResponseTo{
		ID:   mark.ID,
		Name: mark.Name,
	}, nil
}

func (s *MarkService) Update(req dto.MarkUpdateRequestTo) (*dto.MarkResponseTo, error) {
	mark := &entity.Mark{
		Name: req.Name,
		ID:   req.ID,
	}
	err := s.repo.Update(mark)
	if err != nil {
		return nil, errors.New("failed to update mark")
	}
	return &dto.MarkResponseTo{
		ID:   mark.ID,
		Name: mark.Name,
	}, nil
}

func (s *MarkService) Delete(id int64) error {
	err := s.repo.Delete(id)
	if err != nil {
		return err
	}
	return nil
}

func (s *MarkService) GetAll() ([]*dto.MarkResponseTo, error) {
	marks, err := s.repo.GetAll()
	if err != nil {
		return nil, err
	}

	response := make([]*dto.MarkResponseTo, len(marks))
	for i, mark := range marks {
		response[i] = &dto.MarkResponseTo{
			ID:   mark.ID,
			Name: mark.Name,
		}
	}
	return response, nil
}

File: 251002/Belyatskiy/5/lab1 - Copy/publisher/src/services/mark.service.ts
import { MarkRequestTo, MarkResponseTo } from '../dtos/mark.dto.js';
import { Mark } from '../entities/mark.entity.js';
import { validate } from 'class-validator';
import { getRepository } from 'typeorm';
import {cacheService} from "./cache.service.js";

const MARKER_CACHE_PREFIX = 'mark';
const MARKER_FIND_BY_ID_PREFIX = `${MARKER_CACHE_PREFIX}:findById`;
const MARKER_FIND_ALL_PREFIX = `${MARKER_CACHE_PREFIX}:findAll`;

export class MarkService {
    private markRepository = getRepository(Mark);

    async create(markDto: MarkRequestTo): Promise<MarkResponseTo> {
        const errors = await validate(markDto);
        if (errors.length > 0) {
            throw new Error('Validation failed');
        }
        await cacheService.invalidatePattern(`${MARKER_FIND_ALL_PREFIX}*`);
        const mark = this.markRepository.create(markDto);
        await this.markRepository.save(mark);
        return {
            id: mark.id,
            name: mark.name,
        };
    }

    async findAll(page: number = 1, limit: number = 10, sortBy: string = 'id', order: 'ASC' | 'DESC' = 'ASC'): Promise<MarkResponseTo[] | null | undefined> {
        const cacheKey = MARKER_FIND_ALL_PREFIX;
        return await cacheService.getOrSet<MarkResponseTo[]>(cacheKey, async () => {
            const [marks] = await this.markRepository.findAndCount({
                skip: (page - 1) * limit,
                take: limit,
                order: {[sortBy]: order},
            });
            return marks.map(mark => ({
                id: mark.id,
                name: mark.name,
            }));
        })
    }

    async findById(id: number): Promise<MarkResponseTo | undefined |null> {
        const cacheKey = `${MARKER_FIND_BY_ID_PREFIX}:${id}`;
        return await cacheService.getOrSet<MarkResponseTo>(cacheKey, async () => {
            const mark = await this.markRepository.findOne({where: {id}});
            if (mark) {
                return {
                    id: mark.id,
                    name: mark.name,
                };
            }
            return undefined;
        })
    }

    async update(id: number, markDto: MarkRequestTo): Promise<MarkResponseTo | undefined> {
        const errors = await validate(markDto);
        if (errors.length > 0) {
            throw new Error('Validation failed');
        }

        const mark = await this.markRepository.findOne({ where: { id } });
        if (mark) {
            Object.assign(mark, markDto);
            await this.markRepository.save(mark);
            await cacheService.invalidate(`${MARKER_FIND_BY_ID_PREFIX}:${id}`);
            await cacheService.invalidatePattern(`${MARKER_FIND_ALL_PREFIX}*`);
            return {
                id: mark.id,
                name: mark.name,
            };
        }
        return undefined;
    }

    async delete(id: number): Promise<boolean> {
        const result = await this.markRepository.delete(id);
        await cacheService.invalidate(`${MARKER_FIND_BY_ID_PREFIX}:${id}`);
        await cacheService.invalidatePattern(`${MARKER_FIND_ALL_PREFIX}*`);
        return result.affected !== null && result.affected !== undefined && result.affected > 0;
    }
}